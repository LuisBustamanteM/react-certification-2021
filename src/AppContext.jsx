import React, {createContext, useReducer} from 'react'

export function appReducer(state, action) {
    switch (action.type) {
        case 'UPDATE_QUERY': {
            return {
                ...state,
                query: action.value
            }
        }
        case 'TOGGLE_MODE': {
            return {
                ...state,
                darkMode: action.value
            }
        }
        case 'GET_VIDEOS': {
            return {
                ...state,
                videos: action.value
            }
        }
        case 'ADD_FAVORITES': {
            let storedIds = JSON.parse(window.sessionStorage.getItem('favoriteVideos')) || []
            storedIds = [...storedIds, action.value]
            window.sessionStorage.setItem("favoriteVideos", JSON.stringify(storedIds))

            return {
                ...state,
                favoriteIds: [
                    ...state.favoriteIds,
                    action.value
                ]
            }
        }
        case 'REMOVE_FAVORITES': {
            let storedIds = JSON.parse(window.sessionStorage.getItem('favoriteVideos')) || []
            let removedIds = storedIds.filter(id => id !== action.value)
            window.sessionStorage.setItem("favoriteVideos", JSON.stringify(removedIds))

            return {
                ...state,
                favoriteIds: removedIds
            }
        }
        case 'LOGIN': {
            window.sessionStorage.setItem("isLoggedIn", "true")
            window.sessionStorage.setItem("userData", JSON.stringify(action.value))
            return {
                ...state,
                isLoggedIn: true,
                userData: action.value
            }
        }
        case "LOGOUT": {
            window.sessionStorage.removeItem("isLoggedIn")
            window.sessionStorage.removeItem("userData")
            return {
                ...state,
                isLoggedIn: false,
                userData: {}
            }
        }
        default:
            break
    }

    return state
}

export const initialState = {
    darkMode: true,
    query: "Wizeline",
    videos: [],
    favoriteIds: JSON.parse(window.sessionStorage.getItem("favoriteVideos")) || [],
    isLoggedIn: !!window.sessionStorage.getItem("isLoggedIn"),
    userData: JSON.parse(window.sessionStorage.getItem("userData")) || {},
}

export const StateContext = createContext()
export const DispatchContext = createContext()

const AppContext = (props) => {
    const [state, dispatch] = useReducer(appReducer, initialState)

    return (
        <DispatchContext.Provider value={dispatch}>
            <StateContext.Provider value={state}>
                {props.children}
            </StateContext.Provider>
        </DispatchContext.Provider>

    )

}

export default AppContext

