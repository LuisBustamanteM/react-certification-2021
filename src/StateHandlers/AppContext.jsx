import React, {createContext, useReducer} from 'react'
import {appReducer, initialState} from "./AppReducer";

export const StateContext = createContext()
export const DispatchContext = createContext()

const AppContext = (props) => {
    const [state, dispatch] = useReducer(appReducer, initialState)

    return (
        <DispatchContext.Provider value={dispatch}>
            <StateContext.Provider value={state}>
                {props.children}
            </StateContext.Provider>
        </DispatchContext.Provider>

    )

}

export default AppContext

