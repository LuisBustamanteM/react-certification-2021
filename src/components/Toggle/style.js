import {colors} from '../../globalStyle';
import styled from 'styled-components';

export const ToggleContainer = styled.div`
    margin-right: 1em;
    display: flex;
    align-items: center;
`;

export const Switch = styled.label`
  font-size: 1em;
  margin-right: 0.8em;
  color: ${colors.backgroundLight};
;`

export const Mode =  styled.span`
    color: ${colors.backgroundLight};
`;

/* The switch - the box around the slider */
export const ToggleSwitch = styled.div`
    position: relative;
    display: inline-block;
    width: 50px;
    height: 24px;
    margin-right: 20px;
    
    /* Hide default HTML checkbox */
    input {
        opacity: 1;
        width: 0;
        height: 0;
    }
    
    input:checked + .slider{
      background-color: #19857b;
    }

    input:focus + .slider {
      box-shadow: 0 0 1px #19857b;
    }
        
    input:checked + .slider:before {
      transform: translateX(26px);
    }
   
`

/* The slider */
export const ToggleSlider = styled.div`
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: #b8bb21;
    transition: .4s;
    border-radius: 34px;
    
    :before {
        position: absolute;
        content: "";
        height: 18px;
        width: 18px;
        left: 4px;
        bottom: 4px;
        background-color: #093633;
        box-shadow: 2px 2px 2px black;
        transition: .4s;
        border-radius: 50%;
    }
`

