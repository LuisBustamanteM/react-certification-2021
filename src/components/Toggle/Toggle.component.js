import React from 'react';
import styled from 'styled-components';
import { BsToggle2Off } from 'react-icons/bs';
import {ToggleContainer, Switch, Mode} from './style';

const ToggleComponent = () => {
  return (
    <ToggleContainer>
        <Switch>
           <BsToggle2Off/>
        </Switch>
        <Mode>Dark Mode</Mode>
    </ToggleContainer>
  );
};

export default ToggleComponent;
