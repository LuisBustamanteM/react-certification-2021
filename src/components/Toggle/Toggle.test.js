import React from 'react';

import Toggle from './index';
import {render, getByText, getByRole, act, fireEvent} from "@testing-library/react";
import AppContext from "../../AppContext";

describe("Testing <Toggle/>", () => {
    it("Changes mode when click on<Toggle/> ", () => {
        const {container} = render(
            <AppContext>
                <Toggle/>
            </AppContext>
            );


        const toggleLabel = (text) => getByText(container, text)
        const slider = () => getByRole(container, "slider")

        expect(toggleLabel("Dark Mode"))
        expect(slider())

        act(() => {
            fireEvent.click(slider())
        })

        expect(toggleLabel("Light Mode"))


    });

});