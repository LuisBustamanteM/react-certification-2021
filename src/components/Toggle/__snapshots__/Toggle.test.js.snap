// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Toggle/> <Toggle/> matches snapshot 1`] = `
<div>
  <div
    class="sc-bdfBQB kImYyL"
  >
    <label
      class="sc-gsTEea iyoBSX"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 11c.628-.836 1-1.874 1-3a4.978 4.978 0 0 0-1-3h4a3 3 0 1 1 0 6H9z"
        />
        <path
          d="M5 12a4 4 0 1 1 0-8 4 4 0 0 1 0 8zm0 1A5 5 0 1 0 5 3a5 5 0 0 0 0 10z"
        />
      </svg>
    </label>
    <span
      class="sc-dlfnuX GDpQq"
    >
      Dark Mode
    </span>
  </div>
</div>
`;
